openwisp_controller/api/serializers.py

```diff
@@ -0,0 +1,137 @@
+import collections
+import json
+
+from rest_framework import serializers
+from swapper import load_model
+
+from openwisp_users.api.mixins import FilterSerializerByOrgOwned
+from openwisp_utils.api.serializers import ValidatedModelSerializer
+
+Template = load_model('config', 'Template')
+Vpn = load_model('config', 'Vpn')
+Device = load_model('config', 'Device')
+Config = load_model('config', 'Config')
+TemplateTag = load_model('config', 'TemplateTag')
+Organization = load_model('openwisp_users', 'Organization')
+
+
+class BaseMeta:
+    read_only_fields = ['created', 'modified']
+
+
+class TemplateSerializer(ValidatedModelSerializer):
+    config = serializers.JSONField()
+    tags = serializers.StringRelatedField(many=True, read_only=True)
+    default_values = serializers.JSONField()
+    organization = serializers.PrimaryKeyRelatedField(
+        allow_null=False, queryset=Organization.objects.all(), required=True
+    )
+
+    class Meta(BaseMeta):
+        model = Template
+        fields = [
+            'id',
+            'name',
+            'tags',
+            'organization',
+            'type',
+            'backend',
+            'default',
+            'required',
+            'default_values',
+            'config',
+            'created',
+            'modified',
+        ]
+
+
+class VpnSerializer(FilterSerializerByOrgOwned, ValidatedModelSerializer):
+    config = serializers.JSONField()
+
+    class Meta(BaseMeta):
+        model = Vpn
+        fields = '__all__'
+
+
+class DeviceConfigSerializer(serializers.ModelSerializer):
+    config = serializers.JSONField()
+    context = serializers.JSONField()
+
+    class Meta:
+        model = Config
+        fields = ['backend', 'status', 'templates', 'context', 'config']
+
+
+class DeviceListSerializer(serializers.ModelSerializer):
+    config = DeviceConfigSerializer(write_only=True)
+
+    class Meta(BaseMeta):
+        model = Device
+        fields = [
+            'id',
+            'name',
+            'organization',
+            'mac_address',
+            'key',
+            'last_ip',
+            'management_ip',
+            'model',
+            'os',
+            'system',
+            'notes',
+            'config',
+        ]
+
+    def create(self, validated_data):
+        config_data = validated_data.pop('config')
+        device = Device.objects.create(**validated_data)
+        config_templates = config_data.pop('templates')
+        config = Config.objects.create(device=device, **config_data)
+        for template in config_templates:
+            config.templates.add(template.pk)
+        return device
+
+
+class DeviceDetailSerializer(serializers.ModelSerializer):
+    config = DeviceConfigSerializer()
+
+    class Meta(BaseMeta):
+        model = Device
+        fields = DeviceListSerializer.Meta.fields
+
+    def update(self, instance, validated_data):
+        config_data = None
+
+        if self.data['config'] is None and validated_data.get('config'):
```

Comment:
> hey, @atb00ker I am facing one more issue:
> i.e., after a successful PUT/PATCH request my response data remains the same, I am expecting the response data to change it to the updated data.

That's because you are not suppose to access self.data before saving your instance updates.
The data variable is populated on call and is send with response.

If you want to access it before saving changes, try 
`self.initial_data['config.config'] ` instead of `self.data['config']`.

Reply:
@atb00ker with this line I was trying to check if the config of the device instance is already present or not, and then creating the config instance if not present, but now I have worked out a different approach i.e., now I am checking if `backend` field is present or not in that device instance if not then config is not present and then I am creating the config if the request data contains it.

When I tried with this `self.initial_data['config.config']` I was getting error with PATCH requests, because with this we are assuming that the device already contains the config instance.

Reply:
> we are assuming that the device already contains the config instance.

The line was just an example to show that we use `.initial_data()` and not `.data()`
Bottom line is stay away from `.data()` unless you want to manipulate the data you are sending to the user.
Use `.initial_data()` in it's place and adjust your code to accommodate cases of `.initial_data()`, that shouldn't be hard.

Reply:
Yes, @atb00ker Can you please check now? :smiley: